src/lib.rs

```diff
@@ -1360,6 +1365,203 @@ where
     Some(Ratio::new(n1, d1))
 }
 
+#[cfg(not(all(feature = "bigint", feature = "std")))]
+macro_rules! to_primitive_small {
+    ($($type_name:ty)*) => ($(
+        impl ToPrimitive for Ratio<$type_name> {
+            fn to_i64(&self) -> Option<i64> {
+                self.to_integer().to_i64()
+            }
+
+            fn to_u64(&self) -> Option<u64> {
+                self.to_integer().to_u64()
+            }
+
+            fn to_f64(&self) -> Option<f64> {
+                Some(self.numer.to_f64().unwrap() / self.denom.to_f64().unwrap())
+            }
+        }
+    )*)
+}
+
+#[cfg(not(all(feature = "bigint", feature = "std")))]
+to_primitive_small!(u8 i8 u16 i16 u32 i32);
+
+#[cfg(all(feature = "std", not(feature = "bigint")))]
+macro_rules! to_primitive_64 {
+    ($($type_name:ty)*) => ($(
+        impl ToPrimitive for Ratio<$type_name> {
+            fn to_i64(&self) -> Option<i64> {
+                self.to_integer().to_i64()
+            }
+
+            fn to_u64(&self) -> Option<u64> {
+                self.to_integer().to_u64()
+            }
+
+            fn to_f64(&self) -> Option<f64> {
+                Some(ratio_to_f64(
+                    <i128 as From<$type_name>>::from(self.numer),
+                    <i128 as From<$type_name>>::from(self.denom)
+                ))
+            }
+        }
+    )*)
+}
+
+#[cfg(all(feature = "std", not(feature = "bigint")))]
+to_primitive_64!(i64 u64);
+
+#[cfg(all(feature = "bigint", feature = "std"))]
+impl<T: Clone + Integer + ToPrimitive + ToBigInt> ToPrimitive for Ratio<T> {
+    fn to_i64(&self) -> Option<i64> {
+        self.to_integer().to_i64()
+    }
+
+    fn to_i128(&self) -> Option<i128> {
+        self.to_integer().to_i128()
+    }
+
+    fn to_u64(&self) -> Option<u64> {
+        self.to_integer().to_u64()
+    }
+
+    fn to_u128(&self) -> Option<u128> {
+        self.to_integer().to_u128()
+    }
+
+    fn to_f64(&self) -> Option<f64> {
+        let numer: BigInt = self.numer.to_bigint()?;
+        let denom: BigInt = self.denom.to_bigint()?;
+
+        Some(ratio_to_f64(numer, denom))
```

Comment:
How does the performance of the `BigInt` version compare to the `i128` version? It must at least involve the allocator to make `BigInt`s, even when we're within the `MIN`/`MAX_EXACT_INT` fast path otherwise. Maybe it would be worth trying `to_i64` here first, to call `ratio_to_f64::<i128>` when that suffices.